components/layout/wrapper.rs

```diff
@@ -43,49 +43,50 @@ pub type NonOpaqueStyleAndLayoutData = *mut AtomicRefCell<PersistentLayoutData>;
 pub trait LayoutNodeLayoutData {
     /// Similar to borrow_data*, but returns the full PersistentLayoutData rather
     /// than only the style::data::NodeData.
-    fn borrow_layout_data(&self) -> Option<AtomicRef<PersistentLayoutData>>;
-    fn mutate_layout_data(&self) -> Option<AtomicRefMut<PersistentLayoutData>>;
+    fn borrow_layout_data(&self) -> AtomicRef<PersistentLayoutData>;
+    fn mutate_layout_data(&self) -> AtomicRefMut<PersistentLayoutData>;
+    fn has_layout_data(&self) -> bool;
     fn flow_debug_id(self) -> usize;
 }
 
 impl<T: GetLayoutData> LayoutNodeLayoutData for T {
-    fn borrow_layout_data(&self) -> Option<AtomicRef<PersistentLayoutData>> {
-        unsafe {
-            self.get_style_and_layout_data().map(|opaque| {
-                let container = *opaque.ptr as NonOpaqueStyleAndLayoutData;
-                (*container).borrow()
-            })
-        }
+    fn borrow_layout_data(&self) -> AtomicRef<PersistentLayoutData> {
+        unsafe { (*self.ensure_layout_data()).borrow() }
     }
 
-    fn mutate_layout_data(&self) -> Option<AtomicRefMut<PersistentLayoutData>> {
-        unsafe {
-            self.get_style_and_layout_data().map(|opaque| {
-                let container = *opaque.ptr as NonOpaqueStyleAndLayoutData;
-                (*container).borrow_mut()
-            })
-        }
+    fn mutate_layout_data(&self) -> AtomicRefMut<PersistentLayoutData> {
+        unsafe { (*self.ensure_layout_data()).borrow_mut() }
+    }
+
+    fn has_layout_data(&self) -> bool {
+        self.get_style_and_layout_data().is_some()
     }
 
     fn flow_debug_id(self) -> usize {
-        self.borrow_layout_data().map_or(0, |d| d.flow_construction_result.debug_id())
+        self.borrow_layout_data().flow_construction_result.debug_id()
     }
 }
 
 pub trait LayoutNodeHelpers {
-    fn initialize_data(self);
+    fn ensure_layout_data(&self) -> NonOpaqueStyleAndLayoutData;
 }
 
-impl<T: LayoutNode> LayoutNodeHelpers for T {
-    fn initialize_data(self) {
-        if self.borrow_layout_data().is_none() {
-            let ptr: NonOpaqueStyleAndLayoutData =
-                Box::into_raw(box AtomicRefCell::new(PersistentLayoutData::new()));
-            let opaque = OpaqueStyleAndLayoutData {
-                ptr: unsafe { NonZero::new(ptr as *mut AtomicRefCell<PartialPersistentLayoutData>) }
-            };
-            self.init_style_and_layout_data(opaque);
-        }
+impl<T: GetLayoutData> LayoutNodeHelpers for T {
+    fn ensure_layout_data(&self) -> NonOpaqueStyleAndLayoutData {
```

Comment:
So I'm slightly scared about this change in the sense that now calling `borrow_layout_data` can trigger a non thread-safe allocation, so I'll have to check this much more carefully.